### PR TITLE
Implement Date Filter

### PR DESCRIPTION
**Pull request by sbishel**
#### Summary
Implements a Date filter

#### Ticket Link
https://community.mattermost.com/boards/team/qghzt68dq7bopgqamcnziq69ao/b4ctmm917q3rzdfi451qtuxjzcw/vuswsftt1rb873datosmmbtwkpa/cn8kdzqxwsjyqmpf1s7fi3ct96c

Note - This PR is built on top of 
https://github.com/mattermost/focalboard/pull/4287, which has been approved but needs tested as it changes some fundamental things.

### CODE REVIEW

**Comment by wiggin77** [comment]
FYI - I am waiting for the dependent PR to be merged before reviewing.

File: webapp/src/components/viewHeader/dateFilter.tsx [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    const offsetDate = value ? new Date(value.getTime() + timeZoneOffset(value.getTime())) : undefined
+    const [input, setInput] = useState<string>(getDisplayDate(offsetDate))
+
+    const locale = intl.locale.toLowerCase()
+    if (locale && locale !== 'en' && !loadedLocales[locale]) {
+        // eslint-disable-next-line global-require
+        loadedLocales[locale] = require(`moment/locale/${locale}`)
+    }
+
+    // (new Date(date.getFullYear(), date.getMonth(), date.getDate(), 12))
```
Comment (harshilsharma63) [inline webapp/src/components/viewHeader/dateFilter.tsx:99]: nit: can this be removed?

File: webapp/src/components/viewHeader/dateFilter.tsx [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    if (offsetDate) {
+        displayValue = getDisplayDate(offsetDate)
+    }
+
+    let buttonText = displayValue
+    if (!buttonText) {
+        buttonText = intl.formatMessage({id: 'DateFilter.empty', defaultMessage: 'Empty'})
+    }
+
+    const className = ''
```
Comment (harshilsharma63) [inline webapp/src/components/viewHeader/dateFilter.tsx:132]: We don't seem to be writing to this variable. Can this be removed?

File: webapp/src/components/viewHeader/filterValue.tsx [hunk 3]
```diff
@@ -1,8 +1,14 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.
 import React, {useState} from 'react'
+
 import {useIntl} from 'react-intl'
 
+// import DateRange from '../../properties/date/date'
```
Comment (harshilsharma63) [inline webapp/src/components/viewHeader/filterValue.tsx:7]: nit: unused imports

**Review by harshilsharma63 — APPROVED** [review]
LGTM 👍 Works nicely for date ranges as well!

**Review by wiggin77 — COMMENTED** [review]
LGTM :+1:

**Review by ogi-m — COMMENTED** [review]
Tested the following:
- Date property (incl. date range) with is / is before / is after / is set / is not set
- Created time with is / is before / is after
- Last updated time with is / is before / is after

It's working perfectly for the most part, but I'm seeing one issue where the 'is' option doesn't work initially:
1. Open a board and set the date property on a card
2. Apply the date filter with the 'is' option and select the date that was set on the card in step 1
3. Note the card is filtered out even though it should be shown
4. Change the date filter to any other option (e.g. 'is before')
5. Change it back to 'is'
6. Card is now shown as expected, as it should have after step 2

https://user-images.githubusercontent.com/15314183/211533692-5eec198c-ca6d-48ba-8f72-6901ba58b907.mov

**Comment by ogi-m** [comment]
@sbishel I've filed this issue separately if we want to merge first and fix it later - https://github.com/mattermost/focalboard/issues/4445

**Pull request merged**